stdlib/public/core/DiscontiguousSlice.swift

```diff
@@ -0,0 +1,451 @@
+    var c = 0
+    for range in subranges.ranges {
+      c += base[range].count
+    }
+    return c
+  }
+
+  public var isEmpty: Bool {
+    subranges.isEmpty
+  }
+
+  public func distance(from start: Index, to end: Index) -> Int {
+    let ranges = subranges.ranges[start._rangeOffset ... end._rangeOffset]
+    guard ranges.count > 1 else {
+      return _base[ranges[0]].distance(from: start.base, to: end.base)
+    }
+    let firstRange = ranges.first!
+    let lastRange = ranges.last!
+    let head = _base[firstRange].distance(
+      from: start.base, to: firstRange.upperBound)
+    let tail = _base[lastRange].distance(
+      from: lastRange.lowerBound, to: end.base)
+    let middle = ranges[1 ..< (ranges.endIndex - 1)].reduce(0) {
+      $0 + _base[$1].count
+    }
+    return head + middle + tail
+  }
+  
+  public func index(after i: Index) -> Index {
+    // Note: index validation performed by the underlying collections only
+    let currentRange = subranges.ranges[i._rangeOffset]
+    let nextIndex = _base[currentRange].index(after: i.base)
+    if nextIndex < currentRange.upperBound {
+      return Index(_rangeOffset: i._rangeOffset, base: nextIndex)
+    }
+    
+    let nextOffset = i._rangeOffset + 1
+    guard nextOffset < subranges.ranges.endIndex else {
+      return endIndex
+    }
+
+    return Index(
+      _rangeOffset: nextOffset, base: subranges.ranges[nextOffset].lowerBound)
+  }
+
+  public subscript(i: Index) -> Base.Element {
+    // Note: index validation performed by the base collection only
+    _base[subranges.ranges[i._rangeOffset]][i.base]
+  }
+  
+  public subscript(bounds: Range<Index>) -> DiscontiguousSlice<Base> {
+    let baseBounds = bounds.lowerBound.base ..< bounds.upperBound.base
+    let baseSlice = base[baseBounds]
+    let baseAdjustedBounds = baseSlice.startIndex ..< baseSlice.endIndex
+    let subset = subranges.intersection(RangeSet(baseAdjustedBounds))
+    return DiscontiguousSlice<Base>(_base: base, subranges: subset)
+  }
+
+  internal func _index(of baseIndex: Base.Index) -> Index? {
+    let rangeOffset = subranges.ranges
+      ._partitioningIndex { $0.upperBound >= baseIndex }
+    let subrange = subranges.ranges[rangeOffset]
+    guard subrange.contains(baseIndex) else {
+      return nil
+    }
+    return Index(_rangeOffset: rangeOffset, base: baseIndex)
+  }
+
+  public func _customIndexOfEquatableElement(_ element: Element) -> Index?? {
+    var definite = true
+    for (i, range) in subranges.ranges.enumerated() {
+      guard let baseResult = _base[range]
+        ._customIndexOfEquatableElement(element) else {
+        definite = false
+        continue
+      }
+      guard let baseIndex = baseResult else {
+        continue
+      }
+      return Index(_rangeOffset: i, base: baseIndex)
+    }
+    if definite {
+      return .some(nil)
+    } else {
+      return nil
+    }
+  }
+
+  public func _customLastIndexOfEquatableElement(_ element: Element) -> Index?? {
```

Comment:
Contribution from the 80-character line police:

```suggestion
  public func _customLastIndexOfEquatableElement(
    _ element: Element
  ) -> Index?? {
```